state/entrysync.go

```diff
@@ -0,0 +1,313 @@
+package state
+
+import (
+	"sync"
+	"time"
+
+	"github.com/FactomProject/factomd/common/interfaces"
+	"github.com/FactomProject/factomd/common/messages"
+	"github.com/FactomProject/factomd/common/primitives"
+	"github.com/FactomProject/factomd/database/databaseOverlay"
+)
+
+var (
+	// EntrySyncMax is the maximum amount of entries to request concurrently
+	EntrySyncMax = 750 // good-ish value
+	// EntrySyncRetry dictates after what period to retry an unanswered request
+	EntrySyncRetry = time.Second * 10
+	// EntrySyncWarning is the number of failed requests before warning in the console.
+	EntrySyncWarning = 16
+)
+
+// EntrySync is responsible for sending the requests to fetch missing entries to the network.
+// (aka 2nd pass sync)
+// The general algorithms are as follows:
+//
+// Sync:
+// Start with the last known complete height
+// 1. Load the DBlock and iterate over all EBlock hashes
+// 2. For each EBlock, iterate over the Entry hashes
+// 3. If the DB has the entry, skip, otherwise add the hash to the ask queue
+// 4. Collate all missing entry hashes for each EBlock and add it to eblock queue
+// 5. Inform state we are processing up to that eblock
+//
+// Ask:
+// 1. Look at the first N records of the ask queue
+// 2. If the entry exists, remove it from queue
+// 3. If it's the first time or time threshold passed, send a request to the network
+//
+// Check:
+// 1. Look at the first record of the eblock queue
+// 2. Go through all missing records and check if they exist
+// 3. If there are still missing records, wait and go back to 1
+// 4. If the eblock is finished, inform the state we have all the entries up to that eblock
+type EntrySync struct {
+	s *State
+
+	position uint32
+
+	askMtx sync.RWMutex
+	askMap map[[32]byte]bool // entry hash => ask
+	asks   []*entrySyncAsk   // chronological
+
+	ebMtx   sync.Mutex
+	eblocks []*entrySyncEBlock
+
+	closer chan interface{}
+}
+
+// an item in the ask queue
+type entrySyncAsk struct {
+	hash  interfaces.IHash
+	time  time.Time
+	tries int
+}
+
+// an item in the eblock queue
+type entrySyncEBlock struct {
+	height  uint32
+	missing []interfaces.IHash
+}
+
+// NewEntrySync creates a new EntrySync
+func NewEntrySync(s *State) *EntrySync {
+	es := new(EntrySync)
+	es.s = s
+	es.closer = make(chan interface{}, 1)
+	es.askMap = make(map[[32]byte]bool)
+	return es
+}
+
+// Stop initiates shutdown and will stop all associated goroutines
+func (es *EntrySync) Stop() {
+	close(es.closer)
+}
+
+// check routine, for description see EntrySync comment
+func (es *EntrySync) check() {
+	for {
+		select {
+		case <-es.closer:
+			return
+		default:
+		}
+
+		if len(es.eblocks) == 0 { // outside of mutex but should be fine
```

Comment:
Would it be simpler to implement `es.eblocks` as a buffered channel? That way you would not need this polling/wait approach, not need the ebMtx

Reply:
Using a channel means I'd still have to have an internal loop that sleeps until the eblock is finished, so there's no good way to get rid of polling/wait. It would remove some of the overhead, though. 

I pushed a version that implements this, please let me know what you think.

Reply:
Cool, I like the channel use